File: algorithms/Java/Sorting/QuickSort.java
package algorithms.Java.Sorting;

// Quick Sort is a divide-and-conquer algorithm that selects a 'pivot' element from the array and partitions the other elements into two sub-arrays according to whether they are less than or greater than the pivot. The sub-arrays are then sorted recursively.

public class QuickSort {

  public static void quickSort(int[] arr, int low, int high) {
    if (low < high) {
      int pi = partition(arr, low, high);
      quickSort(arr, low, pi - 1);
      quickSort(arr, pi + 1, high);
    }
  }

  public static int partition(int[] arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;
    for (int j = low; j < high; j++) {
      if (arr[j] < pivot) {
        i++;
        // Swap arr[i] and arr[j]
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }
    }
    // Swap arr[i + 1] and arr[high] (or pivot)
    int temp = arr[i + 1];
    arr[i + 1] = arr[high];
    arr[high] = temp;
    return i + 1;
  }

  public static void main(String[] args) {
    int[] arr = { 64, 34, 25, 12, 22, 11, 90 };
    quickSort(arr, 0, arr.length - 1);
    System.out.println("Sorted array:");
    for (int i : arr) {
      System.out.print(i + " ");
    }
  }
}

File: algorithms/Python/Sorting/quick_sort.py
# quick_sort.py

class quick_sort:
    def __init__(self, arr):
        self.arr = arr

    def sort(self):
        if len(self.arr) <= 1:
            return self.arr
        else:
            pivot = self.arr.pop()
            greater = []
            lesser = []
            for i in self.arr:
                if i > pivot:
                    greater.append(i)
                else:
                    lesser.append(i)
            return quick_sort(lesser).sort() + [pivot] + quick_sort(greater).sort()
        
        print(sort([5, 3, 8, 6, 7, 2]).sort())